### DISCUSSION TITLE
Editor support

### DISCUSSION THREAD

**Discussion by CFiggers** [hylang/hy#2358] [General]
What editors are people using with Hy? How's the editor support story?

I mostly use VS Code, and I notice the top downloaded Hy extension in the store is pretty bare bones—pretty much just syntax highlighting and that's it. 

It's common for Lisps in particular to offer REPL integration via editor plugins, so I'm curious whether that exists for Hy and in what editor.

**Comment by Kodiologist** [comment]
Emacs's inferior-Lisp mode works fine for Hy. I added some basic support for Org Babel in [Daylight](https://github.com/Kodiologist/Daylight). I don't think anybody has been serious about maintaining editor support, and it's not a topic that interests me.

**Comment by cmhobbs** [comment]
I like to use it with Geany with the Lisp document type.

**Comment by CFiggers** [comment]
It would have made sense to mention [hylang/hy-mode](https://github.com/hylang/hy-mode) for Emacs and [hylang/vim-hy](https://github.com/hylang/vim-hy) for vim, considering they're "first party" editor plugins hosted under the hylang GitHub organization.

Is there any interest in a hylang/vscode-hy ? And who would be the decision-maker for such a thing? 

If there is interest, I wouldn't mind contributing substantially to it. I already have a simple "Evaluate selected form in REPL" integration patched into my local VS Code and that alone is already a huge ergonomic improvement. 

I would take significant inspiration from [BetterThanTomorrow/calva](https://github.com/BetterThanTomorrow/calva) for Clojure and Janet's [janet-lang/vscode-janet](https://github.com/janet-lang/vscode-janet).

**Reply by Kodiologist** [reply]
I'd be happy to bring CFiggers/vscode-hy into the hylang organization if you're interested in being the primary maintainer.

**Reply by CFiggers** [reply]
Sure, I'm happy to take that on. 

CFiggers/vscode-hy is a fork of [xuqinghan/vscode-hy](https://github.com/xuqinghan/vscode-hy), which hasn't had any commits since 2018. I made a PR just to add REPL integration. If I will be the primary maintainer, there's some other housekeeping stuff I'd want to update.

Which would make more sense? Go ahead and pull into hylang organization as-is and then make changes, or update CFiggers/vscode-hy first?

**Reply by Kodiologist** [reply]
I've created hylang/vscode-hy as a fork of CFiggers/vscode-hy and added you as a maintainer. What happens now is your choice.

**Reply by CFiggers** [reply]
Could you please turn on Issue tracking in hylang/vscode-hy? Since it's a fork, those are off by default. Thanks.

**Reply by Kodiologist** [reply]
I think you can do that.

**Reply by CFiggers** [reply]
I tried before bothering you about it.

**Reply by Kodiologist** [reply]
I upgraded your collaborator role. Can you do it now?

**Reply by CFiggers** [reply]
Yes! Thank you.

**Reply by danieltanfh95** [reply]
Using `vscode-hy` with [bracket-select](https://marketplace.visualstudio.com/items?itemName=chunsen.bracket-select) and [send to terminal](https://marketplace.visualstudio.com/items?itemName=babakks.vscode-send-to-terminal), works pretty well for my setup. Happily deploying Hy in production, would love to see more editor support, I don't mind paying even.

**Reply by CFiggers** [reply]
@danieltanfh95 I just pushed v0.0.3 of [hylang/vscode-hy](http://www.github.com/hylang/vscode-hy), which adds ParEdit-style structural editing commands and on-type auto-formatting. 

v0.0.1 added an "Evaluate in REPL" shortcut and v0.0.2 was an overhaul of the syntax highlighting. 

In the future I want to add Intellisense autocompletion and, if I can figure out how to do it, a background Hy REPL to dynamically pull doc information into a pop-up on symbol mouse-over.

I'd be really interested in your feedback if you give it a try!

**Reply by danieltanfh95** [reply]
> @danieltanfh95 I just pushed v0.0.3 of [hylang/vscode-hy](http://www.github.com/hylang/vscode-hy), which adds ParEdit-style structural editing commands and on-type auto-formatting.
> 
> v0.0.1 added an "Evaluate in REPL" shortcut and v0.0.2 was an overhaul of the syntax highlighting.
> 
> In the future I want to add Intellisense autocompletion and, if I can figure out how to do it, a background Hy REPL to dynamically pull doc information into a pop-up on symbol mouse-over.
> 
> I'd be really interested in your feedback if you give it a try!

Thanks, I'll go test it out!

**Comment by rej696** [comment]
I have been using Neovim with the vim-hy plugin, and `Olical/conjure` plugin for a great integrated repl (works out of the box with hy). I also use `eraserhd/parinfer-rust` neovim plugin which makes lisp editing a breeze!

**Reply by MicahElliott** [reply]
Awesome. Sounds like vscode and vim are pretty first class environments for Hy then. Would love to see at least Emacs be on that plane too. :crossed_fingers:

**Comment by MicahElliott** [comment]
I'm using Emacs, as I'd expect the majority of lispers do. But the tooling is a big hole for adopting Hy. And the lisp interactive repl tooling/workflow IMO is the main reason to even consider Hy.

In order to get completion working (which I think is a must-have), you need to have [jedhy](https://github.com/ekaschalk/jedhy) installed and working. You have to already be down a path to getting [hy-mode](https://github.com/hylang/hy-mode) installed, and then realize you need a 4-year-old Jedhy python package that states clearly:

> _Last Hy version verified working for: Hy 0.16.0_

So then you go track down that Hy release and discover that going back to Python v3.8 isn't far enough back. That's where I gave up and can't justify spending more time just to try out Hy in Emacs to see if it's even got a feasible workflow. Even if jedhy were working, we'd want it to not rely on company since there are newer/simpler completion frameworks now.

So this request is to somehow ensure that Emacs (and maybe other editor) tooling is somewhat up-to-date, even if that means the hylang org does the dev/maintenance.

Clojure has become a huge success in large part due to the existence of CIDER. Hy doesn't need something similarly amazing, but it does need some basics like completion to work.

**Reply by Kodiologist** [reply]
[hy-mode](https://github.com/hylang/hy-mode) is already under the aegis of hylang, but it has no active maintainers. I'm happy to move jedhy to the hylang organization if somebody's interested in maintaining it. The organization has accumulated lots of repositories that people simply stopped maintaining, and it's silly.

**Reply by Kodiologist** [reply]
Wait, no, scratch that, jedhy isn't free software. It would need to be rewritten from scratch, replaced, or given a free license.

**Reply by MicahElliott** [reply]
Thanks for looking into it! Hy has amazing potential with just a bit more editor support, optimizing the OOTB experience. At some point (if no one else has done it yet), I'll plan to try replicating the jedhy functionality to work with [consult](https://github.com/minad/consult) completion. And getting some docs setup working shouldn't be too hard.

Better route though is probably to try to get @ekaschalk to change the license.

**Reply by MicahElliott** [reply]
Filed a ticket for Eric to consider license change: https://github.com/ekaschalk/jedhy/issues/5

**Reply by MicahElliott** [reply]
@Kodiologist I had an email exchange with Eric (jedhy maintainer). I can forward the thread to you if wanted, but in summary his points were:
- not active in open source anymore
- lost access to github account
- not educated on licensing
- if you can fork it, proceed with doing that

So I'd hope that is enough for hylang group to be able to fork, add PSF license, and assume ownership. Not sure if anything more official than this needs to be done. I'm happy to help more with this part if you need.

Then actually getting jedhy updated to work with recent versions can be a secondary step. I'm happy to attempt this but just can't make any promises about completing given the missing context/history/experience.

**Reply by Kodiologist** [reply]
So it sounds like the legal issues can be surmounted. Thank you for going above and beyond looking into this. But I still wouldn't want to put it under hylang without an active maintainer. Why don't you make your own fork (either in the formal sense, on GitHub, or just a local clone on your machine) and see if you can get `jedhy` working to your satisfaction with Hy master? If you do, and you're interested in continuing to maintain `jedhy`, let me know, and have Eric email me the sentence "I agree to license jedhy under the Expat license." (He's the sole author of `jedhy` except for one commit that's too small to be copyrightable.) Then I'll make this happen.

**Reply by Kodiologist** [reply]
Actually, make that "I agree to license jedhy under the GNU General Public License as published by the Free Software Foundation; either version 3, or (at the licensee's option) any later version." `jedhy` looks like it has a lot of code copied from `hy-mode`, which is GPL ≥3.

**Reply by Kodiologist** [reply]
I got the email. `jedhy` is now GPL ≥3, as God intended.

**Reply by pkulev** [reply]
I am refreshing jedhy's code for a couple of days, half of tests are passing now. I think I will get fully working version soon and then start thinking on hy-mode.
https://github.com/ekaschalk/jedhy/compare/master...pkulev:jedhy:master?expand=1

Can upstream of jedhy be forked into hylang organization soon? I'm wondering where to send PRs 😅

**Reply by Kodiologist** [reply]
It still needs a maintainer.

**Reply by pkulev** [reply]
I think I can maintain it, at least for a while. I want to overhaul emacs hy-mode and for this I need jedhy running perfectly.

**Reply by Kodiologist** [reply]
Okay. Once you get jedhy working to your satisfaction with Hy master, if you're still interested in active maintenance, let me know.

**Comment by CFiggers** [comment]
I just pushed v0.0.3 of [vscode-hy](http://www.github.com/hylang/vscode-hy), which adds ParEdit-style structural editing commands and on-type auto-formatting (which mostly means auto-indentation and whitespace management). v0.0.1 added REPL integration and v0.0.2 was an overhaul of the syntax highlighting. 

Testers and contributors welcome.

**Comment by Esnos33** [comment]
Does someone made progress with editor setup? I wish to try it out in emacs

**Comment by atisharma** [comment]
Neovim. I keep trying emacs and keep coming back.
